### PR TITLE
docs: io guide

### PR DESCRIPTION
**Pull request by florian-hoenicke**

### CODE REVIEW

File: chapters/io/index.rst [hunk 1]
```diff
@@ -0,0 +1,315 @@
+How to handle input and output for Flows
```
Comment (cristianmtr) [inline chapters/io/index.rst]:
```suggestion
Input and Output for Flows
```

File: chapters/io/index.rst [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+=======================================
+
+This chapter explains how to provide input data to the `Flow API`. Moreover, it describes how to handle the output via
```
Comment (cristianmtr) [inline chapters/io/index.rst]: Why the line break in the sentence?

File: chapters/io/index.rst [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+callback functions.
+
+Input
+-----
+The input data for the Flow operations `Flow.index`, `Flow.update` and `Flow.search` can be provided in three different ways.
+In the following, the input functionality is shown using the `Flow.index` function. `Flow.update` and `Flow.search` are used analogous.
```
Comment (cristianmtr) [inline chapters/io/index.rst]:
```suggestion
In the following, the input functionality is shown using the `Flow.index` function. `Flow.update` and `Flow.search` are similar.
```

File: chapters/io/index.rst [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+
+#. A single Document can be sent through the Flow as shown below.
+
+    .. highlight:: python
+    .. code-block:: python
+
+        with f:
+            f.index(Document('doc content'))
+
+#. An Iterable of Documents can be provided instead. The following example shows a Generator fed into a Flow.
```
Comment (cristianmtr) [inline chapters/io/index.rst]:
```suggestion
#. An Iterable of Documents can be provided instead. The following example shows a `Generator` fed into a `Flow`.
```

I guess these words should all be in backticks, as they are class / type names

File: chapters/io/index.rst [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+
+    .. highlight:: python
+    .. code-block:: python
+
+        with f:
+            f.index((Document(f'doc content {i}') for i in range(10)))
+
+
+#. The data can also be wrapped in a parameter less function.
```
Comment (cristianmtr) [inline chapters/io/index.rst]:
```suggestion
#. The data can also be wrapped in a parameter-less function.
```

File: chapters/io/index.rst [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+
+        async def async_input_fn():
+            for _ in range(10):
+                yield np.random.random([4])
+                await asyncio.sleep(0.1)
+
+        with AsyncFlow() as f:
+            async f.index(input_fn()):
+
+The `Flow.delete` function accepts Document Ids instead of Documents.
```
Comment (cristianmtr) [inline chapters/io/index.rst]: Lower case ids

File: chapters/io/index.rst [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+            f.delete(input_function_delete)
+
+Special Input Functions
+-----------------------
+
+There are some functions of the Flow API which simplify the input handling:
+`Flow.index_lines`, `Flow.index_ndjson`, `Flow.index_csv`, `Flow.index_files`, `Flow.index_ndarray`
+`Flow.search_lines`, `Flow.search_ndjson`, `Flow.search_csv`, `Flow.search_files` and `Flow.search_ndarray`
+
+The following examples show the `Flow.index_*` functions. Search works analogous.
```
Comment (cristianmtr) [inline chapters/io/index.rst]:
```suggestion
The following examples show the `Flow.index_*` methods. It's similar for `search_*`.
```

File: chapters/io/index.rst [hunk 8]
```diff
@@ -0,0 +1,315 @@
...
+Here, a csv file is used to index Documents. The possible ways of feeding in the csv are shown.
+The function `Flow.index_lines` can be used in combination with `line_format = 'csv'`.
+A simpler version is to use `F.index_csv` where the `line_format` parameter is not needed.
+The csv data can be provided as file handler or directly as `str array`.
+    .. highlight:: csv
```
Comment (cristianmtr) [inline chapters/io/index.rst]: Might need a line break here. Not sure how rst renders this one

File: chapters/io/index.rst [hunk 9]
```diff
@@ -0,0 +1,315 @@
...
+            f.index_lines(open('input.csv'), line_format='csv')
+
+        with f:
+            f.index_csv(open('input.csv').readlines())
+
+        with f:
+            f.index_csv(open('input.csv').readlines(), line_format='csv')
+
+
+Using JSON Lines works analogous.
```
Comment (cristianmtr) [inline chapters/io/index.rst]:
```suggestion
It's similar when using JSON lines.
```

File: chapters/io/index.rst [hunk 10]
```diff
@@ -0,0 +1,315 @@
...
+            # in case of an Exception, the flow execution continues and calls this `on_error` handler
+
+        def handle_search_done():
+            # this handler is always called regardless of Exceptions
+
+        with Flow.load_config(os.path.join(cur_dir, 'flow.yml')) as f:
+            f.search([doc], on_done=handle_response, on_error=handle_error, on_always=handle_search_done)
+
+
+It can be handy to use the built in `print` function as `on_done` callback.
```
Comment (cristianmtr) [inline chapters/io/index.rst]:
```suggestion
It can be handy to use the built-in `print` function as `on_done` callback.

```

File: chapters/io/index.rst [hunk 11]
```diff
@@ -0,0 +1,315 @@
...
+    .. code-block:: python
+
+        with f:
+            f.search(input_fn, on_done=print)
+
+    
+Insights
+--------
+
+When using the `Flow.*` functions, Jina builds and sends Protobuf messages to the relevant pods.
```
Comment (cristianmtr) [inline chapters/io/index.rst]: Uppercase `Pods`

File: chapters/io/index.rst [hunk 12]
```diff
@@ -0,0 +1,315 @@
...
+    }
+  }
+}
+```
+
+The structure of this message is defined in the format of [protobuf](https://docs.jina.ai/chapters/proto/docs.html).
+Check more details of the data structure at [`jina.proto`](https://docs.jina.ai/chapters/proto/docs.html#jina.proto).
+
+`request` contains input data and related metadata.
+The input is a 3*8 matrix that is sent to the Flow, which matches 3 `request.index.docs`,
```
Comment (cristianmtr) [inline chapters/io/index.rst]: I think the `*` is confusing the renderer. Might be the same for rst. Check how it looks

**Review by cristianmtr — CHANGES_REQUESTED** [review]
some comments

File: chapters/io/index.rst (hunk 12 above)
Comment (florian-hoenicke) [reply chapters/io/index.rst]: good notion. Seems like we are lucky. It is rendered like a charm.

File: chapters/io/index.rst (hunk 1 above)
Comment (florian-hoenicke) [reply chapters/io/index.rst]: check

File: chapters/io/index.rst (hunk 2 above)
Comment (florian-hoenicke) [reply chapters/io/index.rst]: I wanted to have less than 120 characters in one line. Now, I made the line break after the first sentence. One line break is ignored by the rendering anyways.

File: chapters/io/index.rst (hunk 3 above)
Comment (florian-hoenicke) [reply chapters/io/index.rst]: Changed it to:
For `Flow.update` and `Flow.search`, the input is provided the same way.

File: chapters/io/index.rst (hunk 4 above)
Comment (florian-hoenicke) [reply chapters/io/index.rst]: Thanks, done

File: chapters/io/index.rst (hunk 5 above)
Comment (florian-hoenicke) [reply chapters/io/index.rst]: thanks

File: chapters/io/index.rst (hunk 6 above)
Comment (florian-hoenicke) [reply chapters/io/index.rst]: check

File: chapters/io/index.rst (hunk 7 above)
Comment (florian-hoenicke) [reply chapters/io/index.rst]: Thanks,
I changed it to:
Providing `Documents` to search for works in the same way using the respective functions.

File: chapters/io/index.rst (hunk 8 above)
Comment (florian-hoenicke) [reply chapters/io/index.rst]: check

File: chapters/io/index.rst (hunk 9 above)
Comment (florian-hoenicke) [reply chapters/io/index.rst]: thanks, changed

File: chapters/io/index.rst (hunk 10 above)
Comment (florian-hoenicke) [reply chapters/io/index.rst]: check

File: chapters/io/index.rst (hunk 11 above)
Comment (florian-hoenicke) [reply chapters/io/index.rst]: check

File: chapters/io/index.rst [hunk 13]
```diff
@@ -0,0 +1,319 @@
...
+This chapter explains how to provide input data to the `Flow API`.
+Moreover, it describes how to handle the output via callback functions.
+
+Input
+-----
+The input data for the `Flow` operations `Flow.index`, `Flow.update` and `Flow.search` can be provided in three different ways.
+In the following, the input functionality is shown using the `Flow.index` function.
+For `Flow.update` and `Flow.search`, the input is provided the same way.
+
+#. A single `Document` can be sent through the `Flow` as shown below.
```
Comment (alexcg1) [inline chapters/io/index.rst]: Should we be using backticks here? We're speaking about the Flow itself, not reference to it in code

File: chapters/io/index.rst [hunk 14]
```diff
@@ -0,0 +1,319 @@
...
+        with f:
+            f.delete(['DOC_ID_1', 'DOC_ID_2', 'DOC_ID_3'])
+
+        def input_function_delete():
+            for i in range(10):
+                yield f'{i}'
+        with f:
+            f.delete(input_function_delete)
+
+Special Input Functions
```
Comment (alexcg1) [inline chapters/io/index.rst]: Capitalization - should be sentence case

File: chapters/io/index.rst [hunk 15]
```diff
@@ -0,0 +1,319 @@
...
+
+There are some functions of the `Flow API` which simplify the input handling:
+`Flow.index_lines`, `Flow.index_ndjson`, `Flow.index_csv`, `Flow.index_files`, `Flow.index_ndarray`
+`Flow.search_lines`, `Flow.search_ndjson`, `Flow.search_csv`, `Flow.search_files` and `Flow.search_ndarray`
+
+The following examples show the usage of the `Flow.index_*` functions.
+Providing `Documents` to search for works in the same way using the respective functions.
+Here, a `CSV` file is used to index `Documents`. The possible ways of feeding in the `CSV` are shown.
+The function `Flow.index_lines` can be used in combination with `line_format = '`CSV`'`.
+A simpler version is to use `F.index_csv` where the `line_format` parameter is not needed.
```
Comment (alexcg1) [inline chapters/io/index.rst]: - `Flow.index_csv()`
- `Flow.index_lines()`
- Put `()` after each method to show it's not an attribute
- Should it be `functions` or `methods`?

File: chapters/io/index.rst [hunk 16]
```diff
@@ -0,0 +1,319 @@
...
+The `CSV` data can be provided as file handler or directly as `str array`.
```
Comment (alexcg1) [inline chapters/io/index.rst]: - `as a file handler`

File: chapters/io/index.rst [hunk 17]
```diff
@@ -0,0 +1,319 @@
...
+
+The `Flow.index_files` function can be used if multiple files have to be fed into the `Flow`.
+
+    .. highlight:: python
+    .. code-block:: python
+
+        with f:
+            f.index_files('*.png', on_done=print)
+
+Using `Flow.index_ndarray` and `Flow.search_ndarray`, numpy arrays can be fed into the `Flow`.
```
Comment (alexcg1) [inline chapters/io/index.rst]: Add `()` to methods

File: chapters/io/index.rst [hunk 18]
```diff
@@ -0,0 +1,319 @@
...
+            # in case of an Exception, the flow execution continues and calls this `on_error` handler
+
+        def handle_search_done():
+            # this handler is always called regardless of Exceptions
+
+        with Flow.load_config(os.path.join(cur_dir, 'flow.yml')) as f:
+            f.search([doc], on_done=handle_response, on_error=handle_error, on_always=handle_search_done)
+
+
+It can be handy to use the built-in `print` function as `on_done` callback.
```
Comment (alexcg1) [inline chapters/io/index.rst]: `s/handy/useful`

File: chapters/io/index.rst [hunk 19]
```diff
@@ -0,0 +1,319 @@
...
+You can also combine two types of data, like:
+
+```python
+def input_fn():
+    for g in all_gif_files:
+        with open(g, 'rb') as fp:
+            yield g.encode() + b'JINA_DELIM' + fp.read()
+```
+
+The `crafter` then can be implemented as:
```
Comment (alexcg1) [inline chapters/io/index.rst]: `can then`

File: chapters/io/index.rst [hunk 20]
```diff
@@ -0,0 +1,319 @@
...
+
+    def craft(self, buffer, *args, **kwargs):
+        file_name, img_raw = buffer.split(b'JINA_DELIM')
+        im = Image.open(io.BytesIO(img_raw))
+        # manipulate the image and file_name here
+        # ...
+
+```
+
+As you can see from the examples above, we can use `buffer` to transfer strings and gif videos.
```
Comment (alexcg1) [inline chapters/io/index.rst]: `s/gif videos/animated gifs`

**Review by alexcg1 — CHANGES_REQUESTED** [review]
- should we use term `functions` or `methods` for stuff under `Flow.`?
- remember `()` if using `Flow.index()`, etc. Otherwise it looks like attribute
- Use backticks if we're referring to code or CLI output. Otherwise use normal text (capitalized according to whether the thing is a core Jina component like Flow, Pea, Pod, etc)

File: chapters/core/guides/index.rst [hunk 21]
```diff
@@ -38,4 +38,5 @@ Further information on any specific method can be obtained in the :doc:`../api_r
    ../../evaluation
    ../../incremental_indexing
    ../../stress/index
+   ../../io/index
```
Comment (FionnD) [inline chapters/core/guides/index.rst]: Should we flatten this to just 
`chapters/input_output.rst`

File: chapters/io/index.rst [hunk 22]
```diff
@@ -0,0 +1,319 @@
+How to handle Input and Output for Flows
+=======================================
+
+This chapter explains how to provide input data to the `Flow API`.
+Moreover, it describes how to handle the output via callback functions.
```
Comment (FionnD) [inline chapters/io/index.rst]: I think you can shorten this to one sentence? 

This chapter explains how to input data via `Flow API`and handle outputs via callback functions.

File: chapters/io/index.rst (hunk 2 above)
Comment (cristianmtr) [reply chapters/io/index.rst]: We shouldn't break paragraphs https://jinaai.slack.com/archives/C018QCARHNG/p1612284457147700

File: chapters/io/index.rst [hunk 23]
```diff
@@ -0,0 +1,319 @@
...
+
+Input
+-----
+The input data for the `Flow` operations `Flow.index`, `Flow.update` and `Flow.search` can be provided in three different ways.
+In the following, the input functionality is shown using the `Flow.index` function.
```
Comment (FionnD) [inline chapters/io/index.rst]: In the following **example**,

File: chapters/io/index.rst [hunk 24]
```diff
@@ -0,0 +1,319 @@
...
+        shape: 8
+        dtype: "float64"
+      }
+    }
+  }
+}
+```
+
+The structure of this message is defined in the format of [protobuf](https://docs.jina.ai/chapters/proto/docs.html).
+Check more details of the data structure at [`jina.proto`](https://docs.jina.ai/chapters/proto/docs.html#jina.proto).
```
Comment (FionnD) [inline chapters/io/index.rst]: Find more details

File: chapters/io/index.rst [hunk 25]
```diff
@@ -0,0 +1,319 @@
...
+
+`request` contains input data and related metadata.
+The input is a 3*8 matrix that is sent to the `Flow`, which matches 3 `request.index.docs`,
+and the `request.index.docs.blog.shape` is 8.
+The vector of the matrix is stored in `request.index.docs.blob`, and the `request.index.docs.blob.dtype` indicates the type of the vector.
+
+
+More Examples
```
Comment (FionnD) [inline chapters/io/index.rst]: Sentence case

**Review by FionnD — CHANGES_REQUESTED** [review]
Good guide! Some very minor issues.

File: chapters/io/index.rst (hunk 25 above)
Comment (florian-hoenicke) [reply chapters/io/index.rst]: done thanks

File: chapters/io/index.rst (hunk 24 above)
Comment (florian-hoenicke) [reply chapters/io/index.rst]: check

File: chapters/io/index.rst (hunk 16 above)
Comment (florian-hoenicke) [reply chapters/io/index.rst]: check

File: chapters/io/index.rst (hunk 17 above)
Comment (florian-hoenicke) [reply chapters/io/index.rst]: I see the concern. Actually, f() would mean that I call it without parameter. Maybe something like f(•) or f(...)?

File: chapters/io/index.rst (hunk 18 above)
Comment (florian-hoenicke) [reply chapters/io/index.rst]: thanks

File: chapters/io/index.rst (hunk 19 above)
Comment (florian-hoenicke) [reply chapters/io/index.rst]: thanks done

File: chapters/io/index.rst (hunk 20 above)
Comment (florian-hoenicke) [reply chapters/io/index.rst]: thanks done

File: chapters/core/guides/index.rst (hunk 21 above)
Comment (florian-hoenicke) [reply chapters/core/guides/index.rst]: done

File: chapters/io/index.rst (hunk 22 above)
Comment (florian-hoenicke) [reply chapters/io/index.rst]: Changed it to:
This chapter explains how input and output data is handled by the `Flow API`.

File: chapters/io/index.rst (hunk 23 above)
Comment (florian-hoenicke) [reply chapters/io/index.rst]: check

File: chapters/io/index.rst (hunk 14 above)
Comment (florian-hoenicke) [reply chapters/io/index.rst]: thanks

File: chapters/io/index.rst (hunk 15 above)
Comment (florian-hoenicke) [reply chapters/io/index.rst]: check, thanks,
it should be functions since it can provide a return value

File: chapters/io/index.rst (hunk 13 above)
Comment (florian-hoenicke) [reply chapters/io/index.rst]: I would say so. I put all Jina terms is backticks.

File: chapters/io.rst [hunk 26]
```diff
@@ -0,0 +1,277 @@
+How to handle Input and Output for Flows
+=======================================
+
+This chapter explains how input and output data is handled by the `Flow API`.
+
+Input
+-----
+The input data for the `Flow` functions `flow.index(...)`, `flow.update(...)` and `flow.search(...)` can be provided in three different ways.
```
Comment (nan-wang) [inline chapters/io.rst]: Note `` `italic` `` is for italic in rst. Please use either `` :term:`parameter ` `` or `` `` `` for verbatim 
```suggestion
The input data for the ``Flow`` functions ``flow.index(...)``, ``flow.update(...)`` and ``flow.search(...)`` can be provided in three different ways.
```

**Review by nan-wang — CHANGES_REQUESTED** [review]

**Review by FionnD — APPROVED** [review]
LGTM, I think the name should be more verbose. Like input_output.rst, but it's up to you.

**Review by cristianmtr — APPROVED** [review]

File: chapters/io.rst (hunk 26 above)
Comment (florian-hoenicke) [reply chapters/io.rst]: thanks - I changed all ` to ``

**Review by nan-wang — APPROVED** [review]
LGTM👍

**Comment by florian-hoenicke** [comment]
@alexcg1 Is there more to address for this guide?
